src/Xamarin.Android.Build.Tasks/Tasks/Desugar.cs

```diff
@@ -0,0 +1,109 @@
+
+		[Output]
+		public string [] OutputJars { get; set; }
+
+		public override bool Execute ()
+		{
+			Log.LogDebugMessage ("Desugar Task");
+			Log.LogDebugMessage ("  JavaPlatformJarPath: ", JavaPlatformJarPath);
+			Log.LogDebugMessage ("  DesugarJarPath: ", DesugarJarPath);
+			Log.LogDebugMessage ("  DesugarExtraArguments: ", DesugarExtraArguments);
+			Log.LogDebugMessage ("  OutputDirectory: ", OutputDirectory);
+			Log.LogDebugMessage ("  InputClassesDirectory: ", InputClassesDirectory);
+			Log.LogDebugTaskItems ("  InputJars: ", InputJars);
+
+			if (!Directory.Exists (OutputDirectory))
+				Directory.CreateDirectory (OutputDirectory);
+
+			return base.Execute ();
+		}
+
+		protected override string GenerateCommandLineCommands ()
+		{
+			//   Running command: C:\Program Files (x86)\Java\jdk1.6.0_20\bin\java.exe
+			//     "-jar" "C:\Program Files (x86)\Android\android-sdk-windows\platform-tools\jill.jar"
+			//     "--output-dex" "bin\classes"
+			//     "-classpath" "C:\Users\Jonathan\Documents\Visual Studio 2010\Projects\AndroidMSBuildTest\AndroidMSBuildTest\obj\Debug\android\bin\mono.android.jar"
+			//     "@C:\Users\Jonathan\AppData\Local\Temp\tmp79c4ac38.tmp"
+
+			//var android_dir = MonoDroid.MonoDroidSdk.GetAndroidProfileDirectory (TargetFrameworkDirectory);
+
+			var cmd = new CommandLineBuilder ();
+
+			// Add the JavaOptions if they are not null
+			// These could be any of the additional options
+			if (!string.IsNullOrEmpty (JavaOptions)) {
+				cmd.AppendSwitch (JavaOptions);
+			}
+
+			// Add the specific -XmxN to override the default heap size for the JVM
+			// N can be in the form of Nm or NGB (e.g 100m or 1GB ) 
+			cmd.AppendSwitchIfNotNull ("-Xmx", JavaMaximumHeapSize);
+
+			cmd.AppendSwitchIfNotNull ("-jar ", DesugarJarPath);
+
+			cmd.AppendSwitch ("--bootclasspath_entry ");
+			cmd.AppendFileNameIfNotNull (JavaPlatformJarPath);
+
+			//cmd.AppendSwitchIfNotNull ("-J-Dfile.encoding=", "UTF8");
+
+			if (!string.IsNullOrEmpty (DesugarExtraArguments))
+				cmd.AppendSwitch (DesugarExtraArguments); // it should contain "--dex".
+
+			if (!string.IsNullOrEmpty (InputClassesDirectory)) {
+				cmd.AppendSwitch ("--input ");
+				cmd.AppendFileNameIfNotNull (InputClassesDirectory);
+				cmd.AppendSwitch ("--output ");
+				cmd.AppendFileNameIfNotNull (Path.Combine (OutputDirectory, "__app_classes__.jar"));
+			}
+
+			var outputs = new List<string> ();
+			var md5 = System.Security.Cryptography.MD5.Create ();
+			foreach (var jar in InputJars) {
+				var output = Path.Combine (OutputDirectory, BitConverter.ToString (md5.ComputeHash (Encoding.UTF8.GetBytes (jar))) + Path.GetFileName (jar));
+				outputs.Add (output);
+				cmd.AppendSwitch ("--input ");
+				cmd.AppendFileNameIfNotNull (jar);
+				cmd.AppendSwitch ("--output ");
+				cmd.AppendFileNameIfNotNull (output);
+			}
+
+			OutputJars = outputs.ToArray ();
+			Log.LogDebugTaskItems ("  OutputJars: ", OutputJars);
+
+			return cmd.ToString ();
```

Comment:
You've forgot to add parameter `--min_sdk_version`. It is important for desugar to know application minimum sdk version (default is 1).

Reply:
Weird, we have a test that does pass without it. Could you elaborate and explain why/how it is important?

Reply:
> There are multiple reasons but the main one is that we need to know the application's minSdk to desugar correctly. Also, there is potentially a classpath issue when dealing with composite builds and multiple app importing the same library module. So we do all the desugaring in the main app project once we know the above information.

https://issuetracker.google.com/issues/67452341#comment2

For example, if minSdk < 19, then desugar replace `try-with-resources` to simple `try..catch`

Reply:
Oh, that sounds correct. Okay, we should work on that. Thanks for the info!